### PR TITLE
added service registry handling

### PR DESCRIPTION
**Pull request by Jamie-BitFlight**

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -1,6 +1,8 @@
 locals {
-  ecs_cluster_name = "${element(split("/",var.ecs_cluster_id),3)}"
+  ecs_cluster_name = "${basename(var.ecs_cluster_id)}"
```
Comment (maartenvanderhoef) [inline main.tf]: Ah that's so much nicer!

File: main.tf [hunk 2]
```diff
@@ -96,13 +98,18 @@ module "alb_handling" {
...
+  tags = "${local.tags}"
 }
 
 ###### CloudWatch Logs
 resource "aws_cloudwatch_log_group" "app" {
   count             = "${var.create ? 1 : 0}"
   name              = "${local.ecs_cluster_name}/${var.name}"
-  retention_in_days = 14
+  retention_in_days = "${var.log_retention_in_days}"
+  kms_key_id        = "${var.cloudwatch_kms_key}"
```
Comment (maartenvanderhoef) [inline main.tf]: Nice addition!

File: modules/ecs_container_definition/main.tf [hunk 3]
```diff
@@ -18,7 +18,73 @@ resource "null_resource" "envvars_as_list_of_maps" {
 }
 
 locals {
-  container_definitions = [{
+  use_port = "${var.container_port == "" ? "without_port" : "with_port" }"
```
Comment (maartenvanderhoef) [inline modules/ecs_container_definition/main.tf]: Is it possible to have an empty portMappings ? If not, maybe can we do something with merge so we don't need two almost similar container definitions ?

File: modules/alb_handling/main.tf [hunk 4]
```diff
@@ -5,12 +5,12 @@ data "aws_lb" "main" {
 
 locals {
   # Validate the record type by looking up the map with valid record types
-  route53_record_type = "${lookup(var.allowed_record_types,var.route53_record_type)}"
+  route53_record_type = "${lookup(var.allowed_record_types,var.route53_record_type, "NONE")}"
```
Comment (jonboulle) [inline modules/alb_handling/main.tf]: doesn't this undo the "validation"?

File: modules/ecs_container_definition/main.tf [hunk 5]
```diff
@@ -30,22 +44,13 @@ locals {
...
-    portMappings = [
-      {
-        containerPort = "${var.container_port}"
-        hostPort      = "${var.host_port}"
-        protocol      = "${var.protocol}"
-      },
-    ]
-
-    healthCheck = "${var.healthcheck}"
+    # privileged             = "${var.privileged}"
```
Comment (jonboulle) [inline modules/ecs_container_definition/main.tf]: ?

File: modules/ecs_service/main.tf [hunk 6]
```diff
@@ -1,20 +1,21 @@
...
+  count = "${var.create && var.lb_attached ? 1 : 0}"
 
   triggers {
     listeners = "${join(",", var.aws_lb_listener_rules)}"
   }
 }
 
 resource "aws_ecs_service" "app_with_lb_awsvpc" {
-  count = "${var.create && local.awsvpc_enabled && local.lb_attached ? 1 : 0}"
+  count = "${var.create && local.awsvpc_enabled &&  var.lb_attached  ? 1 : 0}"
```
Comment (jonboulle) [inline modules/ecs_service/main.tf]: looks like an extra space snuck in before "var"

File: modules/ecs_service/main.tf [hunk 7]
```diff
@@ -45,7 +46,7 @@ resource "aws_ecs_service" "app_with_lb_awsvpc" {
 }
 
 resource "aws_ecs_service" "app_with_lb_spread" {
-  count       = "${var.create && !local.awsvpc_enabled && local.lb_attached && var.with_placement_strategy ? 1 : 0}"
+  count       = "${var.create && !local.awsvpc_enabled &&  var.lb_attached  && var.with_placement_strategy ? 1 : 0}"
```
Comment (jonboulle) [inline modules/ecs_service/main.tf]: extra space as above

File: modules/ecs_service/main.tf [hunk 8]
```diff
@@ -112,8 +113,94 @@ resource "aws_ecs_service" "app_with_lb" {
   depends_on = ["null_resource.aws_lb_listener_rules"]
 }
 
+resource "aws_ecs_service" "app_with_network_lb" {
+  count           = "${var.create && !local.awsvpc_enabled && ! var.lb_attached && var.nlb_attached && var.container_port != "" && !var.with_placement_strategy ? 1 : 0}"
```
Comment (jonboulle) [inline modules/ecs_service/main.tf]: space between ! and var.lb_attached

File: modules/ecs_service/main.tf [hunk 9]
```diff
@@ -1,20 +1,21 @@
 locals {
   awsvpc_enabled = "${length(var.awsvpc_subnets) > 0 ? true : false }"
-  lb_attached    = "${var.lb_attached}"
+
+  use_service_registry = "${var.enable_service_discovery}"
```
Comment (jonboulle) [inline modules/ecs_service/main.tf]: unused?

File: modules/ecs_service/main.tf [hunk 10]
```diff
@@ -112,8 +113,94 @@ resource "aws_ecs_service" "app_with_lb" {
   depends_on = ["null_resource.aws_lb_listener_rules"]
 }
 
+resource "aws_ecs_service" "app_with_network_lb" {
```
Comment (jonboulle) [inline modules/ecs_service/main.tf]: what is the difference between this block and `resource "aws_ecs_service" "app_with_lb" {` above? I can't see it

File: main.tf [hunk 11]
```diff
@@ -247,7 +258,7 @@ module "ecs_service" {
   awsvpc_security_group_ids = "${var.awsvpc_security_group_ids}"
 
   # lb_target_group_arn sets the arn of the target_group the service needs to connect to
-  lb_target_group_arn = "${module.alb_handling.lb_target_group_arn}"
+  lb_target_group_arn = "${var.lb_tg_arn == "" ?  (lookup(var.load_balancing_properties,"lb_tg_arn", "") == "" ? module.alb_handling.lb_target_group_arn : lookup(var.load_balancing_properties,"lb_tg_arn", "")) : var.lb_tg_arn }"
```
Comment (jonboulle) [inline main.tf]: should one of these be network_load_balancing_properties?

File: modules/ecs_task_definition/main.tf [hunk 12]
```diff
@@ -20,7 +20,11 @@ resource "aws_ecs_task_definition" "app" {
   # mimic multiple `volume` statements
   # This WILL break in Terraform 0.12: https://github.com/hashicorp/terraform/issues/14037#issuecomment-361358928
   # but we need something that works before then
-  volume = ["${var.host_path_volumes}"]
+  # volume = ["${var.host_path_volumes}"]
```
Comment (jonboulle) [inline modules/ecs_task_definition/main.tf]: what's this about?

File: modules/ecs_container_definition/main.tf (hunk 3 above)
Comment (Jamie-BitFlight) [reply modules/ecs_container_definition/main.tf]: yeah its possible. I have services that are using it with this map selector in it working.

File: modules/alb_handling/main.tf (hunk 4 above)
Comment (Jamie-BitFlight) [reply modules/alb_handling/main.tf]: true it does

File: modules/ecs_container_definition/main.tf (hunk 5 above)
Comment (Jamie-BitFlight) [reply modules/ecs_container_definition/main.tf]: I'm removing that

**Comment by Jamie-BitFlight** [comment]
I'm doing a new pull for some of these changes

**Pull request closed without merging**